Remotion/Web/ClientScript/Scripts/ListMenu.ts

```diff
@@ -0,0 +1,300 @@
+    public readonly Text: string,
+    public readonly Icon: string,
+    public readonly IconDisabled: string,
+    public readonly RequiredSelection: number,
+    public readonly IsDisabled: boolean,
+    public readonly Href: string,
+    public readonly Target: string,
+    public readonly DiagnosticMetadata: Dictionary<string | boolean>,
+    public readonly DiagnosticMetadataForCommand: Dictionary<string | boolean>)
+  {
+  }
+}
+
+class ListMenu
+{
+  private static _listMenuInfos: Dictionary<ListMenu_MenuInfo> = {};
+  private static _itemClassName: string = 'listMenuItem';
+  private static _itemFocusClassName: string = 'listMenuItemFocus';
+  private static _itemDisabledClassName: string = 'listMenuItemDisabled';
+  private static _requiredSelectionAny: number = 0;
+  private static _requiredSelectionExactlyOne: number = 1;
+  private static _requiredSelectionOneOrMore: number = 2;
+
+  public static Initialize ($listMenu: JQuery): void
+  {
+    ArgumentUtility.CheckNotNullAndTypeIsJQuery("$listMenu", $listMenu);
+
+    $listMenu.keydown (function (event)
+    {
+      ListMenu_OnKeyDown (event, $listMenu);
+    });
+  }
+
+  public static AddMenuInfo (listMenu: HTMLElement, menuInfo: ListMenu_MenuInfo): void
+  {
+    this._listMenuInfos[listMenu.id] = menuInfo;
+  }
+
+  public static Update (listMenu: HTMLElement, getSelectionCount: Nullable<() => number>): void
+  {
+    var menuInfo = this._listMenuInfos[listMenu.id];
+    if (menuInfo == null)
+      return;
+
+    var itemInfos = menuInfo.ItemInfos;
+    var selectionCount = -1;
+    if (getSelectionCount != null)
+      selectionCount = getSelectionCount();
+
+    for (var i = 0; i < itemInfos.length; i++)
+    {
+      var itemInfo = itemInfos[i];
+      var isEnabled = true;
+      if (itemInfo.IsDisabled)
+      {
+        isEnabled = false;
+      }
+      else
+      {
+        if (itemInfo.RequiredSelection == this._requiredSelectionExactlyOne && selectionCount != 1)
+        {
+          isEnabled = false;
+        }
+        if (itemInfo.RequiredSelection == this._requiredSelectionOneOrMore && selectionCount < 1)
+        {
+          isEnabled = false;
+        }
+      }
+      // TODO: nullcheck
+      var item = document.getElementById (itemInfo.ID) as HTMLElement;
+      let anchor = $ (item).children().eq (0)[0] as HTMLAnchorElement;
+      var icon = $ (anchor).children().eq (0)[0] as HTMLImageElement;
+      if (isEnabled)
+      {
+        if (icon != null && icon.nodeType == 1)
+          icon.src = itemInfo.Icon;
+        item.className = this._itemClassName;
+        if (itemInfo.Href != null)
+        {
+          if (itemInfo.Href.toLowerCase().indexOf ('javascript:') >= 0)
+          {
+            anchor.href = '#';
+            anchor.removeAttribute ('target');
+            anchor.setAttribute ('javascript', itemInfo.Href);
+            anchor.removeAttribute ('onclick');
+            // TODO: nullcheck anchor.getAttribute ('javascript')
```

Comment:
this will always have the javascript attribute. a check is only needed if we wouldn't get an exception otherwise. We do have a ton of Javascript that makes assumption on what other javascript code in the same class does or what our Renderer generates. we don't want to add guards for all possible bugs in the rendering of the DOM. a NullReferenceException is okay when our own code has a mismatch. What is important is that the line where the error occurs identifies the problem clearly.

Reply:
> a check is only needed if we wouldn't get an exception otherwise

Well `eval(null)` doesn't throw, it just does nothing/returns `null`. That would make one ugly bug to find.

Reply:
Yes, so it satisfies my condition for "check" -> ticket and todo with ticket number please.